src/containers/ttree.d

```diff
@@ -54,73 +60,103 @@ struct TTree(T, Allocator = Mallocator, bool allowDuplicates = false,
 		private alias AllocatorType = Allocator;
 	}
 	else
-		alias AllocatorType = void*;
+		private alias AllocatorType = void*;
 
-	~this()
+	~this() @trusted
+	{
+		scope(failure) assert(false);
+		clear();
+	}
+
+	/**
+	 * Removes all elements from the tree.
+	 */
+	void clear()
 	{
 		if (root is null)
 			return;
-		deallocateNode(root, allocator);
+		static if (stateSize!Allocator > 0)
+			deallocateNode(root, allocator);
+		else
+			deallocateNode(root, null);
 	}
 
 	debug(EMSI_CONTAINERS) invariant()
 	{
 		assert (root is null || _length != 0);
 	}
 
-
-	alias Value = ContainerStorageType!T;
-
-	/// $(B tree ~= item) operator overload.
+	/**
+	 * $(B tree ~= item) operator overload.
+	 */
 	void opOpAssign(string op)(T value) if (op == "~")
 	{
 		insert(value);
 	}
 
 	/**
-	 * Inserts the given value into the tree.
+	 * Inserts the given value(s) into the tree.
 	 *
-	 * This is not a stable insert. You
-	 * will get strange results if you insert into a tree while iterating over
-	 * it.
+	 * This is not a stable insert. You will get strange results if you insert
+	 * into a tree while iterating over it.
 	 *
-	 * Returns: true if any values were added.
+	 * Params:
+	 *     value = the value to insert
+	 *     overwrite = if `true` the given `value` will replace the first item
+	 *         in the tree that is equivalent (That is greater-than and less-than
+	 *         are both false) to `value`. This is useful in cases where opCmp
+	 *         and opEquals for `T` type have different meanings. For example,
+	 *         if the element type is a circle that has a position and a color,
+	 *         the circle could implement `opCmp` to sort by color, and calling
+	 *         `insert` with `overwrite` set to `true` would allow you to update
+	 *         the position of the circle with a certain color in the tree.
+	 * Returns: the number of values added.
 	 */
-	bool insert(T value)
+	size_t insert(T value, bool overwrite = false) @safe
```

Comment:
Should other containers also have overloads of `insert` that take ranges and return size_t?

Reply:
Putting this off until later: https://github.com/dlang-community/containers/issues/105